corehq/motech/repeaters/models.py

```diff
@@ -909,6 +910,49 @@ def datasource_is_subscribed_to(domain, data_source_id):
         ).exists()
 
 
+def _get_state(self):
+    state = RECORD_PENDING_STATE
+    if self.succeeded and self.cancelled:
+        state = RECORD_EMPTY_STATE
+    elif self.succeeded:
+        state = RECORD_SUCCESS_STATE
+    elif self.cancelled:
+        state = RECORD_CANCELLED_STATE
+    elif self.failure_reason:
+        state = RECORD_FAILURE_STATE
+    return state
+
+
+def set_state(self, value):
+    if value == RECORD_EMPTY_STATE:
+        self.succeeded = True
+        self.cancelled = True
+        self.failure_reason = ""
+    elif value == RECORD_SUCCESS_STATE:
+        self.succeeded = True
+        self.cancelled = False
+        self.failure_reason = ""
+    elif value == RECORD_CANCELLED_STATE:
+        self.succeeded = False
+        self.cancelled = True
+        self.failure_reason = ""
+    elif value == RECORD_FAILURE_STATE:
+        self.succeeded = False
+        self.cancelled = False
+        try:
+            reason = self.failure_reason
```

Comment:
jsonobject's [`JsonProperty` raises `AssertionError`](https://github.com/dimagi/jsonobject/blob/61fed25f1dbe9bf231ca2897d1b80b4dfee615b1/jsonobject/base_properties.pyx#L73) if the record does not have a `failure_reason`. That's annoying. It should raise `AttributeError`, which would allow standard use of `getattr`.